### ISSUE TITLE
Can/should we have an API for re-provisioning DRM certificates?

### ISSUE THREAD

**Issue by baconz** [google/ExoPlayer#8039]
### [REQUIRED] Searched documentation and issues
Indeed, I did.

### [REQUIRED] Question
tl;dr what, if any, are the ill effects of "over-provisioning" DRM certificates? As an extreme example, if we were to provision each time the player started up, would we have any problems besides unnecessary resource usage and added latency?

We see a relatively constant low-level trickle of `MediaDrmStateException`, and we've always chalked them up to hw vendor issues on odd devices, but recently somebody on our dev team ended up in a state where they were consistently getting `MediaDrmStateException` caused by `Failed to get key request: General DRM error` on their Pixel 4 after the Android 11 upgrade. It was consistently reproducible so we dove in.

We tried everything imaginable short of a factory reset, but the device was stuck, it just wouldn't give us the key request. Other apps successfully played DRMed content, just not ours. Then we forced a re-provision in our own code (outside of Exo). Things got unstuck and magically started working.

Now we're trying to figure out if this could help the handful of users who are getting this error on a regular basis, and if it might help the multiple handfuls of users who get other various DRM errors on a regular basis.

Our working plan is to trigger a re-provision based on an allow-list of DRM exceptions, and potentially give users a control "recalibrate their player" that they can use when support tells them to.

If this is something that might help other people and isn't a "bad idea" for some opaque DRM-related reason it might be nice to have an API to perform the reprovision from inside of Exo, perhaps by allowing us to inject an overriden `ProvisioningManagerImpl`.

**Comment by andrewlewis** [comment]
> We tried everything imaginable short of a factory reset, but the device was stuck, it just wouldn't give us the key request. Other apps successfully played DRMed content, just not ours. Then we forced a re-provision in our own code (outside of Exo). Things got unstuck and magically started working.

This makes it sound like a problem with the app rather than an issue affecting how the device is provisioned, and it seems like it would be good to try to diagnose that as far as possible first. Would you be able to share a bug report taken just after reproducing the failure to play DRM content with `dev.exoplayer@gmail.com`?

> tl;dr what, if any, are the ill effects of "over-provisioning" DRM certificates? As an extreme example, if we were to provision each time the player started up, would we have any problems besides unnecessary resource usage and added latency?

@rrfrias, please could you comment on this? Thanks.

**Comment by rrfrias** [comment]
> tl;dr what, if any, are the ill effects of "over-provisioning" DRM certificates? As an extreme example, if we were to provision each time the player started up, would we have any problems besides unnecessary resource usage and added latency?

Re-provisioning will cause any offline licenses to be become unplayable. It is probably not a good idea to re-provision each time the player starts up. It might be worthwhile to re-provision in cases when playback failed.

Each app has its own DRM certificate installed when the it provisions. It is possible that there was an issue with the certificate. In that flow however a NotProvisionedException would be thrown when the session is opened. However you are seeing an exception later when trying to generate a key request so we are not sure whether that is the issue. A bugreport will help clarify.  

Andrew, it might make sense to expose this as an API.

**Comment by andrewlewis** [comment]
Thanks for the information!

@baconz Please could you provide a bug report?

**Comment by baconz** [comment]
@andrewlewis unfortunately we didn't capture one from the p4 when it was in its bad state. Apparently we have an old Nvidia tablet that is behaving the same way. I *think* we'll be able to get our hands on it today.

**Comment by RichardDapice** [comment]
Also having the same issue since 11. I will try to get some logs and things together. Previously we have seen this time to time on some cheaper devices. Since Android R we are seeing this much more frequently on devices like Pixel 3, 3a, 4, and 4a as well as some flagship Samsung devices. Since we are seeing this issue only through analytics I have yet to be able to reproduce it. However, we see the user stuck in this failure indefinitely.

Failed to get key request: General DRM error

**Comment by RichardDapice** [comment]
Trying to get more info for now here is a stacktrace.

```"android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error com.google.android.exoplayer2.drm.DrmSession$DrmSessionException: 
android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error at google.DefaultDrmSession.onError(DefaultDrmSession.java:1) 
at DefaultDrmSession.onKeysError(DefaultDrmSession.java:3) 
at DefaultDrmSession.postKeyRequest(DefaultDrmSession.java:5) 
at doLicense(DefaultDrmSession.java:11) 
at acquire(DefaultDrmSession.java:8) 
at acquireSession(DefaultDrmSessionManager.java:17) 
at com.google.android.exoplayer2.source.SampleMetadataQueue.onFormatResult(SampleMetadataQueue.java:13) 
at com.google.android.exoplayer2.source.SampleMetadataQueue.read(SampleMetadataQueue.java:22) 
at com.google.android.exoplayer2.source.SampleQueue.read(SampleQueue.java:2) 
at ChunkSampleStream.readData(ChunkSampleStream.java:3) 
at com.google.android.exoplayer2.BaseRenderer.readSource(BaseRenderer.java:1) 
at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.readToFlagsOnlyBuffer(MediaCodecRenderer.java:3) 
at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:6) 
at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:14) 
at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:17) 
at android.os.Handler.dispatchMessage(Handler.java:102) at android.os.Looper.loop(Looper.java:223) 
at android.os.HandlerThread.run(HandlerThread.java:67) Caused by: android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error at android.media.MediaDrm.getKeyRequest(Native Method) 
at com.google.android.exoplayer2.drm.FrameworkMediaDrm.getKeyRequest(FrameworkMediaDrm.java:5) 
at DefaultDrmSession.postKeyRequest(DefaultDrmSession.java:2) ... 15 more android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error 
at android.media.MediaDrm.getKeyRequest(Native Method) 
at com.google.android.exoplayer2.drm.FrameworkMediaDrm.getKeyRequest(FrameworkMediaDrm.java:5) 
at DefaultDrmSession.postKeyRequest(DefaultDrmSession.java:2) 
at DefaultDrmSession.doLicense(DefaultDrmSession.java:11) 
at DefaultDrmSession.acquire(DefaultDrmSession.java:8) 
at .DefaultDrmSessionManager.acquireSession(DefaultDrmSessionManager.java:17) 
at com.google.android.exoplayer2.source.SampleMetadataQueue.onFormatResult(SampleMetadataQueue.java:13) 
at com.google.android.exoplayer2.source.SampleMetadataQueue.read(SampleMetadataQueue.java:22) 
at com.google.android.exoplayer2.source.SampleQueue.read(SampleQueue.java:2) 
at ChunkSampleStream.readData(ChunkSampleStream.java:3) 
at com.google.android.exoplayer2.BaseRenderer.readSource(BaseRenderer.java:1) 
at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.readToFlagsOnlyBuffer(MediaCodecRenderer.java:3) 
at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:6) 
at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:14)
at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:17)
at android.os.Handler.dispatchMessage(Handler.java:102) at android.os.Looper.loop(Looper.java:223) 
at android.os.HandlerThread.run(HandlerThread.java:67)
```

**Comment by baconz** [comment]
For a variety of pandemic and scheduling reasons I wasn't able to get the tablet yesterday. I'm hoping to be able to grab it over the weekend. I will try to get a proper bug report early next week.

I realize it is not super useful, but this is what we see in the logs just before the error on the Pixel:

```
2020-10-01 08:45:28.106 20411-20592/com.philo.philo.google D/CryptoHal: found drm@1.0 ICryptoFactory clearkey
2020-10-01 08:45:28.107 20411-20592/com.philo.philo.google D/CryptoHal: found drm@1.0 ICryptoFactory default
2020-10-01 08:45:28.108 20411-20592/com.philo.philo.google D/CryptoHal: found drm@1.0 ICryptoFactory widevine
2020-10-01 08:45:28.109 20411-20592/com.philo.philo.google D/CryptoHal: found drm@1.1 ICryptoFactory clearkey
2020-10-01 08:45:28.110 20411-20592/com.philo.philo.google D/CryptoHal: found drm@1.1 ICryptoFactory widevine
2020-10-01 08:45:28.125 20411-20489/com.philo.philo.google W/lo.philo.googl: Accessing hidden method Landroid/net/Uri;-><init>()V (greylist, reflection, allowed)
2020-10-01 08:45:28.127 20411-20592/com.philo.philo.google E/MediaDrm-JNI: Illegal state exception: Failed to get key request: General DRM error (-2000)
```

@RichardDapice I'm not sure if you're on 2.12.0 yet, but we added code to our `ExoMediaDrm.Provider` that knows how to do the reprovisioning (until an Exo API is available):

```
class FrameworkMediaCryptoProvider constructor(
    private val okHttpClient: OkHttpClient
) : ExoMediaDrm.Provider {
    private val mediaDrms: MutableMap<UUID, ExoMediaDrm> = mutableMapOf()

    override fun acquireExoMediaDrm(uuid: UUID): ExoMediaDrm {
        return mediaDrms.getOrPut(uuid) {
            try {
                FrameworkMediaDrm.newInstance(uuid)
            } catch (e: Exception) {
                // various error handling
                DummyExoMediaDrm()
            }
        }
    }

    private fun frameworkMediaDrm(uuid: UUID): FrameworkMediaDrm? {
        return mediaDrms[uuid].let {
            when (it) {
                is FrameworkMediaDrm -> it
                else -> null
            }
        }
    }

    fun provideForceProvisionObservable(uuid: UUID): Observable<Boolean> {
        return frameworkMediaDrm(uuid)?.let { drm ->
            val drmRequest = drm.provisionRequest
            val url: String = drmRequest.defaultUrl + "&signedRequest=" + String(drmRequest.data)
            val okHttpRequest = Request.Builder()
                .url(url)
                .post(ByteArray(0).toRequestBody())
                .build()
            Observable
                .fromCallable { okHttpClient.newCall(okHttpRequest).execute() }
                .subscribeOn(Schedulers.io())
                .map { response ->
                    if (response.isSuccessful) {
                        drm.provideProvisionResponse(response.body!!.bytes())
                        true
                    } else {
                        false
                    }
                }
                .onErrorReturn { false }
        } ?: Observable.just(false)
    }
```

**Comment by rrfrias** [comment]
Thanks for trying to get us a bugreport. The error code in the exception is generic** so a bugreport will be helpful in figuring out the root cause.

** we are looking to correct this in a future release.

**Comment by andrewlewis** [comment]
@baconz Were you able to capture a bug report? Thanks.

**Comment by baconz** [comment]
Sadly, no @andrewlewis. The device I thought we could repro on turned out to be running Widevine CDM version 1.0, and thus I believe it was suffering from a separate issue.

I will keep my eyes peeled, and let you know if we are able to repro again. I've also asked our support team to watch for this issue and get us a bug report if they find it in the wild.

**Comment by andrewlewis** [comment]
Thanks for trying. I guess we shouldn't add an API for reprovisioning (or automatic reprovisioning) until we have a bug report and can check what the underlying cause is. Therefore I will close this for now. Feel free to reopen this or file a new issue if there is a bug report or if there is a strong case for adding an API anyway.

**Comment by RichardDapice** [comment]
I sent a bug report but it looks like the email was rejected. I will try to
find where the format is specified.

**Comment by andrewlewis** [comment]
@RichardDapice Did you put "Issue #8039" in the subject line ([instructions](https://github.com/google/ExoPlayer/blob/release-v2/.github/ISSUE_TEMPLATE/bug.md#required-a-full-bug-report-captured-from-the-device))?

**Comment by RichardDapice** [comment]
Apologies, I didn't put "Issue #". I sent it over.

**Comment by rrfrias** [comment]
Thanks for the bugreport. I took a look but it does not seem to capture the MediaDrmStateException. I do not see any calls to provision or make use of provisioning data. Is it possible to send another bugreport after encountering the problem?

**Comment by rrfrias** [comment]
@baconz let us know if you are able to capture a bugreport with the exception.

**Comment by baconz** [comment]
@rrfrias our support team has confirmed that they have found a few cases in the wild. They are reaching to users today to get bug reports.

**Comment by baconz** [comment]
@rrfrias we caught one in the wild! I have the bug report. Is there an email address that I should can send it to?

**Comment by baconz** [comment]
This looks relevant:

```
11-23 13:34:00.529 media   944   997 E WVCdm   : [crypto_session.cpp(94):MapOEMCryptoResult] Mapping OEMCrypto result: crypto_session_method = PrepareAndSignLicenseRequest, result = 10085
11-23 13:34:00.529 media   944   997 E WVCdm   : [cdm_engine.cpp(298):GenerateKeyRequest] Key request generation failed, status = 323
11-23 13:34:00.529 media   944   997 W WVCdm   : Returns UNKNOWN error for legacy status: 323
11-23 13:34:00.529 media   944   997 W WVCdm   : Returns UNKNOWN error for legacy status: 323
11-23 13:34:00.530 media   944   944 I WVCdm   : [cdm_engine.cpp(756):QuerySessionStatus] Querying session status: session_id = sid10
```

**Comment by andrewlewis** [comment]
> @rrfrias we caught one in the wild! I have the bug report. Is there an email address that I should can send it to?

@baconz Please could you send it to dev.exoplayer@gmail.com with subject 'Issue #8039' (I think @rrfrias has access)? Thanks!

**Comment by baconz** [comment]
Done @andrewlewis

**Comment by rrfrias** [comment]
Thanks for the bugreport. It appear to be a problem with the SoC crypto library. This is a problem we are investigating. We would like your permission to share this bugreport with Qualcomm. Do let us know if you are willing to grant us permission.

> 11-23 13:34:00.529 media   944   997 E WVCdm   : [crypto_session.cpp(94):MapOEMCryptoResult] Mapping OEMCrypto result: crypto_session_method = PrepareAndSignLicenseRequest, result = 10085
> 11-23 13:34:00.529 media   944   997 E WVCdm   : [cdm_engine.cpp(298):GenerateKeyRequest] Key request generation failed, status = 323

**Comment by baconz** [comment]
@rrfrias yes, that is fine with me.

If we're getting Qualcomm involved, I'm assuming it might be a little while before there's a fix up. Would it make sense to move forward with a re-provisioning API in the mean time?

**Comment by ojw28** [comment]
@baconz - Just to clarify, this would only be for convenience, right? Application code would still trigger the API manually, and its benefit is just that application developers wouldn't need to implement re-provisioning themselves. I don't think we want to start re-provisioning automatically in the library, because (a) I don't think we have a way of being certain that re-provisioning is the correct thing to do based on the exceptions that are being thrown, and (b) re-provisioning is a destructive operation for apps that have acquired offline licenses. Does that make sense?

Assigning to @icbaker since it falls under possible DRM improvements.

**Comment by baconz** [comment]
@ojw28 for sure, it would have to manually triggered. I was just thinking it would be nice for the library to provide a convenient and standardized way to do it.

We don't do any offline licensing, and plan to reprovision whenever the user receives `DrmSessionException`, which it a big hammer, but it feels like it is worth it.

**Comment by navjotbedi** [comment]
Any ETA by when we can have this enhancement?

**Comment by kcrg** [comment]
Pls fix

**Comment by PawKanarek** [comment]
One of my custome have exact the same problem. 
Google Pixel 3, Android 11, 

ExoPlayer for Xamarin 2.8.8.0 
```

  --- End of managed Com.Google.Android.Exoplayer2.Drm.DrmSessionDrmSessionException stack trace ---
com.google.android.exoplayer2.drm.DrmSession$DrmSessionException: android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error
	at com.google.android.exoplayer2.drm.DefaultDrmSession.onError(DefaultDrmSession.java:449)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.onKeysError(DefaultDrmSession.java:444)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.postKeyRequest(DefaultDrmSession.java:397)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.doLicense(DefaultDrmSession.java:324)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.acquire(DefaultDrmSession.java:159)
	at com.google.android.exoplayer2.drm.DefaultDrmSessionManager.acquireSession(DefaultDrmSessionManager.java:552)
	at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.onInputFormatChanged(MediaCodecRenderer.java:889)
	at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.onInputFormatChanged(MediaCodecAudioRenderer.java:352)
	at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:566)
	at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:518)
	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:301)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:223)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error
	at android.media.MediaDrm.getKeyRequest(Native Method)
	at com.google.android.exoplayer2.drm.FrameworkMediaDrm.getKeyRequest(FrameworkMediaDrm.java:152)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.postKeyRequest(DefaultDrmSession.java:393)
	... 11 more
	```

**Comment by J6ey** [comment]
The inability to re-provision is especially an issue in different regions of the world like China. Phones made by their vendors are unable to access the correct provisioning URL for DRMs e.g., Widevine, hence the exception.

**Comment by icbaker** [comment]
> The inability to re-provision is especially an issue in different regions of the world like China. Phones made by their vendors are unable to access the correct provisioning URL for DRMs e.g., Widevine, hence the exception.

I don't think the enhancement being discussed here would help in this case.

What we're discussing here is a way to force re-provisioning even after provisioning seems to have completed successfully (because the app is seeing DRM playback failures, despite apparently being provisioned, and it seems that in some cases re-provisioning can resolve them).

If the provisioning URL isn't reachable (as in your case) then surely provisioning will never succeed, so you never need to force a re-provision. If provisioning has somehow succeeded (e.g. when the device had an unrestricted internet connection for some reason), then it seems you **wouldn't** want to force it to re-provision because it would likely fail every time and then your device would be unprovisioned.

If what you're really asking is "can I override the provisioning server URL?" then you can do this by providing a custom `MediaDrmCallback` implementation. More details: https://github.com/google/ExoPlayer/issues/8165#issuecomment-722337202

**Comment by tpiDev** [comment]
Hi,

In previous comment it has been mentioned that bug-report has been shared with _Qualcomm_. Is there any update on it?
Currently i am facing the same issue: **android.media.MediaDrm.error_neg_2000** with one of installed app on **Pixel 4XL** (latest Android 11 / RQ1A.210205.005).

As the application uninstall and install again does not help here (nor OS update), this issue might be really annoying to customers. any ideas how to move on ? Re-provisioning on **MediaDrmStateException** seems to be overkill especially when handling also offline licenses.

**Comment by icbaker** [comment]
@rrfrias Can you comment on @tpiDev's question?

**Comment by rrfrias** [comment]
Sorry for the delay. I missed this earlier. 

@tpiDev, can you reproduce the issue, capture and send a bugreport to dev.exoplayer@gmail.com. I do not have enough information to tell whether this is the same issue as mentioned earlier. Qualcomm has not had an update for this issue. Will try to get their attention.

> Re-provisioning on MediaDrmStateException seems to be overkill especially when handling also offline licenses.

Yes this is unfortunate. Reprovisioning will cause existing offline licensing to become unusable. baconz's log however showed this error when requesting a license (online).

**Comment by PawKanarek** [comment]
Any update on this issue? We discovered 3 devices that are completely unable to play any DRM content on Android 11.

**Comment by rrfrias** [comment]
Missed this earlier. @PawKanarek can you let us know which devices these are. If you can share a bugreport with  dev.exoplayer@gmail.com we can analyze the failure and follow up, possibly with OEMs.

**Comment by PawKanarek** [comment]
@rrfrias Thats the stacktrace from our case
<details>

```
  End of managed Com.Google.Android.Exoplayer2.Drm.DrmSessionDrmSessionException stack trace ---
com.google.android.exoplayer2.drm.DrmSession$DrmSessionException: android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error
	at com.google.android.exoplayer2.drm.DefaultDrmSession.onError(DefaultDrmSession.java:486)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.onKeysError(DefaultDrmSession.java:481)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.postKeyRequest(DefaultDrmSession.java:432)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.doLicense(DefaultDrmSession.java:367)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.acquire(DefaultDrmSession.java:271)
	at com.google.android.exoplayer2.drm.DefaultDrmSessionManager.acquireSession(DefaultDrmSessionManager.java:536)
	at com.google.android.exoplayer2.source.SampleQueue.onFormatResult(SampleQueue.java:798)
	at com.google.android.exoplayer2.source.SampleQueue.readSampleMetadata(SampleQueue.java:566)
	at com.google.android.exoplayer2.source.SampleQueue.read(SampleQueue.java:330)
	at com.google.android.exoplayer2.source.chunk.ChunkSampleStream.readData(ChunkSampleStream.java:379)
	at com.google.android.exoplayer2.BaseRenderer.readSource(BaseRenderer.java:379)
	at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.readToFlagsOnlyBuffer(MediaCodecRenderer.java:812)
	at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:712)
	at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:599)
	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:329)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:247)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: android.media.MediaDrm$MediaDrmStateException: Failed to get key request: General DRM error
	at android.media.MediaDrm.getKeyRequest(Native Method)
	at com.google.android.exoplayer2.drm.FrameworkMediaDrm.getKeyRequest(FrameworkMediaDrm.java:165)
	at com.google.android.exoplayer2.drm.DefaultDrmSession.postKeyRequest(DefaultDrmSession.java:428)
	... 15 more
```

</details>




Devices: 
![image](https://user-images.githubusercontent.com/15096514/148745036-41dab5e8-2a23-488b-8d2e-dc1a4e2a7e0e.png)

And this bug happens only on Android 11.

**Comment by rrfrias** [comment]
@PawKanarek  Is it possible to get a bugreport for the failure? There is not enough information in the stacktrace above to determine a specific failure.

**Comment by PawKanarek** [comment]
@rrfrias Sorry, we don't have access to these devices. This exception occurs on our customers' devices, and what we get is anonymized error report with exception stacktrace, so we cannot directly ask them about full bugreport.

On our devices bug never occurred, but we could observe it sometimes on AppCenter Test Cloud https://appcenter.ms/, but again what we get is only stacktrace from the exception.

**Comment by giovannyvelezalt** [comment]
We have the same issue and stacktrace from:

- OnePlus AC2001 (Android 11)
- OnePlus HD1901 (Android 11)
- LG LM-G850 (Android 11)
- Xiaomi tablet MiTV-AXSO0  (Android 9)
- Xiaomi Redmi Note 8  (Android 10)
- Vivo 1917  (Android 11)
- Vivo 1938 (Android 10)

**Comment by rrfrias** [comment]
@giovannyvelezalt a stacktrace from an android 12 device might be helpful. A bugreport from any android device would definitely be help us investigate this (I do understand that it is hard to access to devices that reproduce the problem)

**Comment by sridhard** [comment]
Any update on this. we are also facing this issue on multiple customer devices

**Comment by giovannyvelezalt** [comment]
> @giovannyvelezalt a stacktrace from an android 12 device might be helpful. A bugreport from any android device would definitely be help us investigate this (I do understand that it is hard to access to devices that reproduce the problem)

@rrfrias well... we haven't been able to reproduce it and when trying to get logs from users we couldn't get  stacktraces of interest. We have the issue open until we get something useful.

**Comment by icbaker** [comment]
It's possible to manually trigger re-provisioning by using the framework `MediaDrm` APIs directly combined with ExoPlayer's `HttpMediaDrmCallback` to hide away lots of the HTTP-related boiler-plate - something like:

```java
HttpMediaDrmCallback drmCallback =
    new HttpMediaDrmCallback(/* defaultLicenseUrl= */ null, new DefaultHttpDataSource.Factory());
try (MediaDrm mediaDrm = new MediaDrm(C.WIDEVINE_UUID)) {
  ProvisionRequest provisionRequest = mediaDrm.getProvisionRequest();
  mediaDrm.provideProvisionResponse(
      drmCallback.executeProvisionRequest(
          C.WIDEVINE_UUID,
          new ExoMediaDrm.ProvisionRequest(
              provisionRequest.getData(), provisionRequest.getDefaultUrl())));
}
```

**Comment by malbanese** [comment]
We also faced problems with users not being able to playback widevine / drm protected content on our app. We've only been able to observe it happening via issues raised through our customer care team, and haven't seen it on any dev devices.

The symptoms were the same - not being able to playback protected content in our app, other apps work fine, and all of the usual steps like reinstalling didn't work for the users. Plenty of other users had identical devices with no playback issues as well.

Anyways, as a test, we built in the ability for us to remotely allow users with the problem to initiate DRM reprovisioning and it's worked to resolve the issue for multiple users at this point.

It does feel like something that should be automatic though.
